benchmarks/cpp/transformer.cpp

```diff
@@ -0,0 +1,194 @@
+
+using namespace nvfuser;
+
+namespace {
+// Note: We test on smaller model and input sizes to avoid high error
+// accumulation for validation.
+static constexpr int64_t B = 2, E = 768, H = 16, S = 128;
+// Note: Dropout probabilities are set to 0. Since the dropout mask is sharded
+// it throws off the seed offset between the sharded nvFuser program and the
+// unsharded reference.
+static constexpr double kDropoutProb = 0.0, kSdpaProb = 0.0, kSdpaScale = 1e-3;
+// Note parameters scaled by kParamScale following weight initialization
+// recommendations:
+// https://huggingface.co/docs/transformers/en/model_doc/gpt2#transformers.GPT2Config.initializer_range
+static constexpr double kParamScale = 0.02;
+} // namespace
+
+// Return reduction tensor view and output of reduction
+static void setupTransformerForward(Fusion* fusion, DataType dtype) {
+  Communicator* communicator_ = &Communicator::getInstance(); // nick TODO call Communicator::getInstance().cleanup() somewhere before program exit
+
+  const int64_t D = communicator_->size(); // number of devices
+
+  std::unique_ptr<DistributedTransformer> model = std::make_unique<DistributedTransformer>(
+        D, B, E, H, S, kDropoutProb, kSdpaProb);
+
+  model->setupForward(fusion, dtype, /*sequence_parallel*/false);
+}
+
+static std::vector<at::Tensor> reference_mlp(
+    at::Tensor x,
+    at::Tensor w0,
+    at::Tensor b0,
+    at::Tensor w1,
+    at::Tensor b1) {
+  auto at_dtype = w0.dtype();
+  auto linear0 = at::linear(x, w0, b0);
+  auto gelu = at::gelu(linear0.to(at::kFloat), "tanh").to(at_dtype);
+  auto linear1 = at::linear(gelu, w1, b1).to(at::kFloat);
+  auto [dropout, mask] = at::native_dropout(linear1, kDropoutProb, true);
+  return {linear0, gelu, linear1, dropout, mask};
+}
+
+static std::vector<at::Tensor> reference_mha(
+    at::Tensor x,
+    at::Tensor w0,
+    at::Tensor b0,
+    at::Tensor w1,
+    at::Tensor b1) {
+  auto linear0 = at::linear(x, w0, b0);
+  auto qkv = linear0.view({B, S, 3 * E}).split(E, 2);
+  for (auto i = 0; i < 3; i++) {
+    qkv[i] = qkv[i].reshape({B, S, H, E / H}).transpose(1, 2);
+  }
+  auto sdpa_out = at::_scaled_dot_product_flash_attention(
+      qkv[0], qkv[1], qkv[2], kSdpaProb, true, false, kSdpaScale);
+  auto sdpa = std::get<0>(sdpa_out);
+  // Reassemble heads (B, H, S, E/H) to (B, S, H, E/H) to (B, S, E)
+  auto y = sdpa.transpose(1, 2).reshape({B * S, E});
+  auto linear1 = at::linear(y, w1, b1).to(at::kFloat);
+  auto [dropout, mask] = at::native_dropout(linear1, kDropoutProb, true);
+  return {linear0, sdpa, linear1, dropout, mask};
+}
+
+static at::Tensor transformerShardTensor_Mesh(
+    at::Tensor tensor,
+    const int64_t axis,
+    const DeviceMesh& mesh,
+    Communicator* communicator_) {
+  const auto device_id = communicator_->deviceId();
+  return nvfuser::shardTensor(tensor, axis, mesh, device_id);
+}
+
+static at::Tensor transformerShardTensor(at::Tensor tensor, TensorView* tv, Communicator* communicator_) {
```

Comment:
Is this function used? It seems hard to use because TensorViews are hidden behind setupForward which is behind runBenchmarkIterations...